src/evict/evict_page.c

```diff
@@ -110,6 +111,21 @@ __wt_evict(WT_SESSION_IMPL *session, WT_REF *ref, uint8_t previous_state, uint32
     if (tree_dead)
         LF_SET(WT_EVICT_CALL_NO_SPLIT);
 
+    if (session != conn->default_session) {
+        /*
+         * Before we enter the eviction generation we want to make sure that this session has put
+         * the history store dhandle in its session cache. If we do not do this, we might end up in
+         * a deadlock waiting to obtain read lock on the dhandle list while another thread waits to
+         * drain generation while holding the write lock on dhandle list. The simplest way to make
+         * sure that this session's dhandle cache has a reference into history store dhandle is
+         * opening/closing a history store cursor. This should not be expensive as cursors are
+         * cached.
+         */
+        session_flags = 0; /* [-Werror=maybe-uninitialized] */
+        WT_RET(__wt_hs_cursor(session, &session_flags, &is_owner));
+        WT_RET(__wt_hs_cursor_close(session, session_flags, is_owner));
```

Comment:
Does this totally solve the issue, or just make it less likely? I didn't think having a reference in the cache was enough to block out exclusive access if nothing was using it?

Reply:
I think this resolves it entirely. Since we have a reference in the cache, eviction after this won't be blocked on dhandle read lock to get dhandle into session cache from global list.
We are not avoiding exclusive access, instead, we are avoiding the need to take that read lock later for eviction.

Does that explain what we are trying? I can explain in person later if needed.

Reply:
@agorrod 
This is a good question, and I don't know the answer -- is it possible to lose a cached session handle by the time we get to the bottom of reconciliation?